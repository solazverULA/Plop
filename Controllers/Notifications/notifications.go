package notifications

import (
	"net/http"
	"encoding/json"
	"github.com/gorilla/mux"
	"log"
	"strconv"
	"../../Models/ModelNotifications"
	"../../Models/ModelListeners"
	//"io/ioutil"
	"mime/multipart"
)

func RegisterNotification(w http.ResponseWriter, r * http.Request) {
	
	vars := mux.Vars(r)
	user_id := vars["id"]
	w.Header().Set("Content-Type", "text/html; charsed-utf-8")
	w.Header().Set("Access-Control-Allow-Origin", "*")

	userid, _ := strconv.Atoi(user_id)

	ids, notification, file, handle, fileExpandible, handleExpandible := GetNotificationRequest(r) //Obtiene los datos del listener
	
	responsenotification := modelnotifications.ResponseNotifications{"success", modelnotifications.CreateNotifications(notification,ids, userid, file, handle, fileExpandible, handleExpandible), "Notificación registrada con éxito"}

	json.NewEncoder(w).Encode(responsenotification)
}

//Función para extraer del request el json de la notificacion
func GetNotificationRequest(r * http.Request) (modellisteners.ListenerVector, modelnotifications.Notifications , multipart.File ,*multipart.FileHeader, multipart.File, *multipart.FileHeader) {
	var ids modellisteners.ListenerVector
	var notification modelnotifications.Notifications
	//f := make(map[string]interface{})
	file, handle, err := r.FormFile("image")
    if err != nil {
        log.Println( "no hay imagen %v", err)
        file=nil
        handle=nil
    }
    fileExpandible, handleExpandible, err2 := r.FormFile("imageExpandible")
    if err2 != nil {
        log.Println( "no hay imagen %v", err2)
        fileExpandible=nil
        handleExpandible=nil
    }
    log.Println( "%v", file)
    notification.Title = r.FormValue("Title")
    notification.Body = r.FormValue("Body")
    //notification.Namebutton = r.FormValue("Namebutton")
    //notification.Action = r.FormValue("Action")
    notification.Type,_ = strconv.Atoi(r.FormValue("Type"))
 	//notification.Shendule = r.FormValue("Shendule")
 	//notification.Created_at = r.FormValue("Created_at")
    lenght, _ := strconv.Atoi(r.FormValue("lenghtListener"))
    for i := 0; i < lenght; i++ {
    	ids.Ids = append(ids.Ids, r.FormValue("Listener" +strconv.Itoa(i) ))
    	log.Println(ids.Ids[i])
    }

	return ids,notification, file, handle, fileExpandible, handleExpandible
}

//Función para enviar notificaciones a ios y android
func SendNotification(w http.ResponseWriter, r * http.Request) {
	vars := mux.Vars(r)		//Obtenemos los valores de la url
	notification_id := vars["idnotification"]
	w.Header().Set("Content-Type", "text/html; charset-utf-8")
	w.Header().Set("Access-Control-Allow-Origin", "*")
	response := modelnotifications.ResponseNotifications{"succes", modelnotifications.SendNotification(notification_id), "Notificación enviada con éxito"}

	//conexion con json
	json.NewEncoder(w).Encode(response)
}

//Función para obtener las notificaciones por usuario
func GetNotificationUser(w http.ResponseWriter, r * http.Request) {
	vars := mux.Vars(r)
	notificationid := vars["id"]

	w.Header().Set("Content-Type", "text/html; charsed-utf-8")
	w.Header().Set("Access-Control-Allow-Origin", "*")

	json.NewEncoder(w).Encode(modelnotifications.GetNotificationUser(notificationid))

}

//Funcion para ver una notificacion
func GetNotification(w http.ResponseWriter, r * http.Request) {
	vars := mux.Vars(r)
	notificationid := vars["id"]

	w.Header().Set("Content-Type", "text/html; charsed-utf-8")
	w.Header().Set("Access-Control-Allow-Origin", "*")

	json.NewEncoder(w).Encode(modelnotifications.GetNotification(notificationid))
}

//Funcion para obtener los datos de listener_hasnotification
func GetListenerHasNotificacionRequest(r * http.Request) modelnotifications.ListenersReceiveNotifications {
	var listener_hasnotification modelnotifications.ListenersReceiveNotifications

	decoder := json.NewDecoder(r.Body)
	err := decoder.Decode(&listener_hasnotification)

	if err != nil {
		log.Fatal(err)
	}

	return listener_hasnotification
}



