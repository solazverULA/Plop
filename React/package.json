{
  "name": "nud-react",
  "version": "1.0.0",
  "private": true,
  "homepage": "https://timcreative.github.io/now-ui-dashboard-react",
  "dependencies": {
    "animate.css": "^3.6.1",
    "babel-preset-env": "^1.7.0",
    "babel-preset-react": "^6.24.1",
    "bootstrap": "4.0.0-beta",
    "chart.js": "^2.7.2",
    "history": "4.7.2",
    "jquery": "3.3.1",
    "node-sass": "^4.9.0",
    "node-sass-chokidar": "^1.3.0",
    "npm-run-all": "4.1.1",
    "perfect-scrollbar": "1.3.0",
    "react": "16.1.0",
    "react-animated-css": "^1.0.4",
    "react-chartjs-2": "2.7.0",
    "react-dom": "16.1.0",
    "react-google-maps": "9.2.2",
    "react-notification-alert": "0.0.7",
    "react-router-dom": "4.2.2",
    "react-scripts": "^1.1.4",
    "reactstrap": "5.0.0-alpha.4"
  },
  "devDependencies": {
    "babel-jest": "^22.0.4",
    "jest": "^22.0.4",
    "react-test-renderer": "^16.0.0"
  },
  "scripts": {
    "build-css": "node-sass-chokidar --include-path ./src --include-path ./node_modules src/ -o src/",
    "watch-css": "npm run build-css && node-sass-chokidar --include-path ./src --include-path ./node_modules src/ -o src/ --watch --recursive",
    "start-js": "react-scripts start",
    "start": "npm-run-all -p watch-css start-js",
    "build": "npm run build-css && react-scripts build",
    "test": "jest",
    "jest-update-snapshot": "jest --updateSnapshot",
    "eject": "react-scripts eject"
  }
}
