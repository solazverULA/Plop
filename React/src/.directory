[Dolphin]
Timestamp=2018,6,27,3,57,52
Version=3
