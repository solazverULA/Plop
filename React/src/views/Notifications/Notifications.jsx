import React from 'react';

/*
import {
    UncontrolledDropdown, DropdownToggle, DropdownMenu, DropdownItem, Table,
} from 'reactstrap';
*/

import {
    Button,  Row,  Modal,  ModalHeader,  ModalBody,  ModalFooter,  Input,  InputGroup, FormGroup, Table, Col
} from 'reactstrap';

/*
// react plugin used to create charts
import { Line, Bar } from 'react-chartjs-2';
*/

// function that returns a color based on an interval of numbers

import {
    PanelHeader,
    Checkbox
} from '../../components';

import Api from '../../api/Api/Api'
import language from "../../api/translator/translator"
import Cookies from 'react-cookies';
class Notifications extends React.Component{

  constructor(props) {
    super(props);
    let Groups =  Cookies.load("userGroups");
    this.state = {visible: false,
                  toggle: false,
                  notifications:[],
                  notificationType:0,
                  User:Cookies.load('userId'),
                  Groups: Groups ? Groups : [],
                  contacts:[],
                  Listeners:[],
                 };

    this.toggleModal = this.toggleModal.bind(this);
    this.registerNotification = this.registerNotification.bind(this);
    this.changeNotificationType = this.changeNotificationType.bind(this);
    Api._getNotificationsUser(this.state.User.Id, (data)=>this.setState({notifications:data ? data : []}));
    Api._getListenerForUserId(this.state.User.Id, (data)=>{this.setState({contacts:data ? data : []})});

  }

  showChart() {
    return (
      <PanelHeader
          size="sm"
      />
    )
  }

  toggleModal() {
    this.setState({
      toggle: !this.state.toggle,
      showAddContacts:false
    });
  }

  registerNotification() {

    if(!this.state.showAddContacts){
      this.setState({showAddContacts:true})
      return;
    }
    Api._RegisterNotification(this.state, this.state.File, this.state.User.Id,this.toggleModal)
  }

  deleteNotification(index) {
    this.setState({notifications:this.state.notifications.filter((data,i)=>i!==index)})
  }

  changeNotificationType(event){
    this.setState({notificationType:Number(event.target.value), Type:Number(event.target.value)})
  }

  addListenerToNotification(contacts) {
    this.state.Listeners.push(contacts.Id)
  }
  addListenerFromGroup(index) {
    this.state.Groups[index].Listener.forEach((data)=>{
      this.addListenerToNotification(data)
    })
  }

  render(){
     return (
            <div>

              {this.showChart()}

              <Modal isOpen={this.state.toggle} toggle={this.toggleModal}
                className={this.props.className}>
                <ModalHeader toggle={this.toggleModal}>{language("CrearNotificacion")}</ModalHeader>
                <ModalBody style={{'paddingLeft':'7%', 'paddingRight':'7%'}}>
                  { !this.state.showAddContacts ? 
                    <div>
                        <InputGroup style={{'marginLeft':'1%'}}>
                        <Input onChange={this.changeNotificationType} style={{'borderRadius':'5px', 'borderColor':'#979797'}} className='business-title-color' type="select" id="notification_type" name="Type" required={true}>
                          <option value="0">{language("NotificacionSimple")}</option>
                          <option value="2">{language("NotificacionURL")}</option>
                        </Input>
                      </InputGroup>

                      <div style={{'paddingBottom':'5%', 'paddingTop':'5%'}}>
                        <div style={{'width':'100%', 'height':'1px', 'backgroundColor':'#616A6B'}}/>
                      </div>

                      { this.state.notificationType === 0?

                      <div style={{'paddingLeft':'7%', 'paddingRight':'7%'}}>
                        <Row>
                          <FormGroup>
                            <div>
                              {language("CreateTitulo")}
                            </div>
                            <InputGroup>
                              <Input style={{width:'300px'}} type="text" id="notification_title" name="Title" placeholder="Notification title" onChange={(event)=>this.setState({Title:event.target.value})} required={true}/>
                            </InputGroup>
                          </FormGroup>
                        </Row>
                        <Row>
                          <FormGroup>
                            {language("AdjuntarImagen")}
                            <InputGroup>
                              <Button style={{width:'150px'}} color="primaryBlue" htmlFor="notification_file" onClick={()=>{document.getElementById("notification_file").click()}}>
                                <i className='now-ui-icons design_image'></i>
                              </Button>
                              <Input onChange={(event)=>{this.setState({File: event.target.files[0]})}} style={{paddingLeft:'100%'}} className='custom-file-upload' type="file" id="notification_file" name="File"/>
                            </InputGroup>
                          </FormGroup>
                        </Row>
                        <Row>
                          <FormGroup>
                            <div className='business-title-color'>
                              {language("CreateMensaje")}
                            </div>
                            <InputGroup>
                              <textarea onChange={(event)=>this.setState({Body:event.target.value})} className='business-title-color' style={{'borderRadius':'5px', 'borderColor':'#979797', 'height':'107px'}} type="text" id="notification_Body" name="Body" required={true}/>
                            </InputGroup>
                          </FormGroup>
                        </Row>
                      </div>
                      : null }

                      { this.state.notificationType === 2 ?

                        <div style={{'paddingLeft':'7%', 'paddingRight':'7%'}}>

      										<Col xs="12" md="6" sm="6" lg="6">
                            <Row>
                              <FormGroup>
                                <div>
                                  {language("CreateTitulo")}
                                </div>
                                <InputGroup>
                                  <Input style={{width:'300px'}} type="text" id="notification_title" name="Title" placeholder="Notification title" onChange={(event)=>this.setState({Title:event.target.value})} required={true}/>
                                </InputGroup>
                              </FormGroup>
                            </Row>

      											<Row style={{'paddingBottom':'2%'}}>
      												<strong>{language("EnlaceNotificacion")}</strong>
      											</Row>

      											<Row>
      												<FormGroup>
      													<div className='business-title-color'>
      														{language("EtiquetaBoton")}
      													</div>
                                <InputGroup>
                                  <Input style={{width:'300px'}} type="text" id="notification_title" name="Title" placeholder="Notification title" onChange={(event)=>this.setState({Namebutton:event.target.value})} required={true}/>
                                </InputGroup>
      												</FormGroup>
      											</Row>
      											<Row>
    													<FormGroup>
      													<div className='business-title-color'>
      														URL
      													</div>
                                <InputGroup>
                                  <Input style={{width:'300px'}} type="text" id="notification_title" name="Title" placeholder="Notification title" onChange={(event)=>this.setState({Action:event.target.value})} required={true}/>
                                </InputGroup>
      												</FormGroup>
      											</Row>
      										</Col>
      										<Col xs="12" md="6" sm="6" lg="6">
                              <FormGroup>
                                <div className='business-title-color'>
                                  {language("CreateMensaje")}
                                </div>
                                <InputGroup>
                                  <textarea onChange={(event)=>this.setState({Body:event.target.value})} className='business-title-color' style={{'borderRadius':'5px', 'borderColor':'#979797', 'height':'107px'}} type="text" id="notification_Body" name="Body" required={true}/>
                                </InputGroup>
                              </FormGroup>
      										</Col>
      									</div>

                      : null }

                    </div>
                  :<div>
                       <Table hover responsive>
                        <thead className="thead-default">
                          <tr>
                          
                          </tr>
                        </thead>
                        <tbody>
                          { this.state.Groups.map((group, i)=>{
                            return(
                              <tr key = {i}>
                                <td className="justify-content-center text-center">{group.Name}</td>
                                <td className="justify-content-center text-center">{group.Listener.length}</td>
                            
                                <td className="justify-content-center text-center">
                                  <Checkbox
                                      label=""
                                      inputProps={{defaultChecked: false, onChange:()=>this.addListenerFromGroup(i)}}

                                  />
                                </td>
                              </tr>
                            );

                          })}
                
                          { this.state.contacts.map((contacts, i)=>{
                            return(
                              <tr key = {i}>
                                <td className="justify-content-center text-center">
                                  {contacts.Nameprofile}
                                </td>
                                <td className="justify-content-center text-center">
                                  {contacts.Gender}
                                </td>
                                <td className="justify-content-center text-center">
                                 <Checkbox
                                      label=""
                                      inputProps={{defaultChecked: false, onChange:()=>this.addListenerToNotification(contacts)}}

                                  />
                                </td>
                              </tr>
                            );

                          })}
                        </tbody>
                      </Table>
                     
                    </div>
                    
                  }

                </ModalBody>
                <ModalFooter className='justify-content-right text-right'>
                 
                  <Button color="primaryBlue" onClick={this.registerNotification}>{language("BotonCrear")}</Button>
                </ModalFooter>

              </Modal>

              { this.state.notifications[0] != null ?

                <Table hover responsive>
                  <thead className="thead-default">
                    <tr>
                      <th className="justify-content-center text-center">{language("CreateTitulo")}</th>
                      <th className="justify-content-center text-center">{language("CreateMensaje")}</th>
                      <th className="justify-content-center text-center">{language("CreateImagen")}</th>
                      <th className="justify-content-center text-center">{language("BotonEliminar")}</th>
                    </tr>
                  </thead>
                  <tbody>
                    { this.state.notifications.map((notifications, i)=>{
                      return(
                        <tr key = {i}>
                          <td className="justify-content-center text-center">
                            {notifications.Notifications.Title}
                          </td>
                          <td className="justify-content-center text-center">
                            {notifications.Notifications.Body}
                          </td>
                          <td className="justify-content-center text-center">
                            <img width="50" height="50" src={"https://drive.google.com/uc?export=view&id="+notifications.Notifications.Srcimage}/>
                          </td>
                          <td className="justify-content-center text-center">
                             <Button style={{width:'50px', alignItems:"center"}} color="primaryBlue" htmlFor="notification_file" onClick={()=>this.deleteNotification(i)}>
                              <i className='now-ui-icons ui-1_simple-remove'></i>
                            </Button>
                          </td>
                        </tr>
                      );

                    })}
                  </tbody>
                </Table>

              :

              <div className="content flex-center animated animatedFadeInUp fadeInUp">
                  <h2>{language("CrearNuevaNotificacion")}</h2>
              </div>

              }

              <div style={{paddingBottom:'1%', paddingRight:'1%'}} className="fixed-right justify-content-right text-right">
                <Button color="primaryBlue" style={{width:'50px', height:'50px'}} className="btn-round btn-icon" onClick={this.toggleModal}>
                  <i className="now-ui-icons ui-1_simple-add"></i>
                </Button>
              </div>

            </div>
        );
    }
}

export default Notifications;
