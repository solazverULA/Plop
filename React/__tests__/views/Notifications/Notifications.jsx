import React from 'react';
import Notifications from '../../../src/views/Notifications/Notifications';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const loading = renderer.create(
    <Notifications/>
  ).toJSON();
  
  expect(loading).toMatchSnapshot();
});
