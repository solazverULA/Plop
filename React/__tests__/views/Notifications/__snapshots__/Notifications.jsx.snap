// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="panel-header panel-header-sm"
  />
  <div
    style={
      Object {
        "marginLeft": "1%",
      }
    }
  >
    <button
      className="btn-round btn-icon btn btn-primaryBlue"
      onClick={[Function]}
      type="button"
    >
      <i
        className="now-ui-icons ui-1_simple-add"
      />
    </button>
  </div>
  <div
    className="content flex-center animated animatedFadeInUp fadeInUp"
  >
    <h2>
      Create a new notification!
    </h2>
  </div>
</div>
`;
